### PR TITLE
[PAE-108] Re organize classes 

### PR DESCRIPTION
**Pull request by andrey-canon**
## Description
This PR **doesn't** add any new functionalities or fix any kind of error, it is just move some classes to its own file in order to improve the code comprehension.
Your comments or suggestion are welcome.

### CODE REVIEW

**Review by diegomillan — APPROVED** [review]
I'm approving this PR, but since we are moving enrollment controllers to another file we should think about changing "ExternalEnrollmentFactory.get_enrollment_controller" to support getting "SalesforceEnrollment" as well. That way our interface would be cleaner and always relying on a method to decide the required controller instance at execution. @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
+"""External enrollments file."""
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Isn't better if each the enrollments types is in its own file? @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+from openedx_external_enrollments.models import (
+    EnrollmentRequestLog,
+    ProgramSalesforceEnrollment,
+)
+
+LOG = logging.getLogger(__name__)
+
+
+class BaseExternalEnrollment(object):
+    """
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Docstrings throughout the file... @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+        try:
+            response = self._execute_post(
+                url=url,
+                headers=self._get_enrollment_headers(),
+                json_data=json_data,
+            )
+        except Exception as error:
+            LOG.error("Failed to complete enrollment. Reason: " + str(error))
+            log_details["response"] = {"error": "Failed to complete enrollment. Reason: " + str(error)}
+            EnrollmentRequestLog.objects.create(request_type=str(self), details=log_details)
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: What should happen if .create fails? @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+        except Exception as error:
+            LOG.error("Failed to get token: " + str(error))
+        else:
+            if response.ok:
+                headers = {
+                    "Accept": "application/json",
+                    "Content-Type": "application/json",
+                    "Authorization": "{} {}".format(
+                        response.json()["token_type"],
+                        response.json()["access_token"]
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Safe indexing. Trailing comma. @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+                    )
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Trailing comma. @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+        return course
+
+    def _get_enrollment_headers(self):
+        token = self._get_auth_token()
+        return {
+            "Content-Type": "application/json",
+            "Authorization": "{} {}".format(
+                token.get("token_type"),
+                token.get("access_token"),
+            )
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Trailing comma. @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+        return user
+
+    def _get_salesforce_data(self, data):
+        """
+
+        :param data:
+        :return:
+        """
+        salesforce_data = {}
+        order_lines = data.get("supported_lines")
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Object of type 'NoneType' has no len() if it takes the default value. @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+                headers=self._get_enrollment_headers(),
+                json_data=json_data,
+            )
+        except Exception as error:
+            LOG.error("Failed to complete enrollment. Reason: " + str(error))
+            log_details["response"] = {"error": "Failed to complete enrollment. Reason: " + str(error)}
+            EnrollmentRequestLog.objects.create(request_type=str(self), details=log_details)
+            return str(error), status.HTTP_400_BAD_REQUEST
+        else:
+            LOG.info('External enrollment response for [%s] -- %s', self.__str__(), response.json())
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Double quotes. @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+            json=json_data,
+        )
+        return response
+
+    def _post_enrollment(self, data, course_settings=None):
+        """
+        """
+        url = self._get_enrollment_url(course_settings)
+        json_data = self._get_enrollment_data(data, course_settings)
+        LOG.info('calling enrollment for [%s] with data: %s', self.__str__(), json_data)
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Double quotes. @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+
+        :param data:
+        :param order_lines:
+        :return:
+        """
+        program_of_interest = {}
+        program = data.get("program")
+        try:
+            email = order_lines[0].get("user_email")
+            openedx_user, _ = get_user(email=email)
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Unused variable _ @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+            )
+            program_of_interest["Secondary_Source"] = data.get(
+                "utm_campaign",
+                program_of_interest.get("Secondary_Source", "Undefined"),
+            )
+            program_of_interest["Tertiary_Source"] = data.get(
+                "utm_medium",
+                program_of_interest.get("Tertiary_Source", "Undefined"),
+            )
+        except:
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Some kind of logging info? @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+            "Secondary_Source",
+            "Tertiary_Source",
+            "Drupal_ID",
+            "Purchase_Type",
+            "PaymentAmount",
+            "Amount_Currency",
+            "Course_Data",
+        ]
+        payload = {
+            "enrollment": {}
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Trailing comma. @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+
+        salesforce_data = self._get_salesforce_data(data)
+        payload["enrollment"].update(salesforce_data)
+
+        payload["enrollment"]["Course_Data"] = self._get_courses_data(
+            data,
+            data.get("supported_lines"),
+        )
+
+        unwanted = set(payload["enrollment"]) - set(valid_keys)
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Safe checker for "enrollement". @andrey-canon

File: openedx_external_enrollments/external_enrollments.py [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+        for unwanted_key in unwanted:
+            del payload["enrollment"][unwanted_key]
```
Comment (Squirrel18) [inline openedx_external_enrollments/external_enrollments.py]: Could a key error occur with the del statement? @andrey-canon

File: openedx_external_enrollments/factory.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
+"""Openedx external enrollments factory file."""
+from openedx_external_enrollments.external_enrollments import (
+    EdxEnterpriseExternalEnrollment,
+    EdxInstanceExternalEnrollment,
+)
+
+
+class ExternalEnrollmentFactory:
+    """
```
Comment (Squirrel18) [inline openedx_external_enrollments/factory.py:11]: Docstring. @andrey-canon

File: openedx_external_enrollments/factory.py [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+    """
+    @classmethod
+    def get_enrollment_controller(cls, controller):
+        """
+        Return the an instance of the enrollment controller.
+        """
+        if controller.lower() == 'openedx':
+            return EdxInstanceExternalEnrollment()
+        else:
```
Comment (Squirrel18) [inline openedx_external_enrollments/factory.py:20]: Unnecessary else statement. @andrey-canon

File: openedx_external_enrollments/tests/tests_factory.py [hunk 17]
```diff
@@ -0,0 +1,29 @@
+"""Tests factory file."""
```
Comment (Squirrel18) [inline openedx_external_enrollments/tests/tests_factory.py:1]: Is this the same test? Did you change something? @andrey-canon

**Review by Squirrel18 — CHANGES_REQUESTED** [review]

File: openedx_external_enrollments/external_enrollments.py (hunk 1 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: do you think it is better a folder called external enrollments and a individual file for each class ?

File: openedx_external_enrollments/tests/tests_factory.py (hunk 17 above)
Comment (andrey-canon) [reply openedx_external_enrollments/tests/tests_factory.py:1]: The code is the same, the only thing that has changed is the import paths

File: openedx_external_enrollments/external_enrollments.py (hunk 1 above)
Comment (Squirrel18) [reply openedx_external_enrollments/external_enrollments.py]: Definitely. @andrey-canon

File: openedx_external_enrollments/tests/tests_factory.py (hunk 17 above)
Comment (Squirrel18) [reply openedx_external_enrollments/tests/tests_factory.py:1]: Ok.

File: openedx_external_enrollments/external_enrollments.py (hunk 2 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 3 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 4 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 5 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 6 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 7 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 8 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 9 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 10 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 11 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 12 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 13 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/external_enrollments.py (hunk 14 above)
Comment (andrey-canon) [reply openedx_external_enrollments/external_enrollments.py]: out of the scope of this pr

File: openedx_external_enrollments/factory.py (hunk 15 above)
Comment (andrey-canon) [reply openedx_external_enrollments/factory.py:11]: out of the scope of this pr

File: openedx_external_enrollments/factory.py (hunk 16 above)
Comment (andrey-canon) [reply openedx_external_enrollments/factory.py:20]: out of the scope of this pr

**Comment by andrey-canon** [comment]
@Squirrel18 addressed

**Comment by Squirrel18** [comment]
I approved the organization of the code in its own file, but not the code style and we should address my comments in a future PR.

**Review by Squirrel18 — APPROVED** [review]

**Pull request merged**